nixos/modules/services/databases/openldap.nix

```diff
@@ -24,20 +23,79 @@ let
       include ${cfg.rootpwFile}
     ''}
     directory ${cfg.dataDir}
-    ${cfg.extraDatabaseConfig}
+    ${if cfg.extraDatabaseConfig != null then cfg.extraDatabaseConfig else ""}
   '');
-  configOpts = if cfg.configDir == null then "-f ${configFile}"
-               else "-F ${cfg.configDir}";
-in
 
-{
-
-  ###### interface
+  configDir = if cfg.configDir != null then cfg.configDir else "/etc/openldap/slapd.d";
 
+  ldapValueType = let
+    singleLdapValueType = types.either types.str (types.submodule {
+      options = {
+        path = mkOption {
+          type = types.path;
+          description = ''
+            A path containing the LDAP attribute. This is included at run-time, so
+            is recommended for storing secrets.
+          '';
+        };
+      };
+    });
+  in types.either singleLdapValueType (types.listOf singleLdapValueType);
+
+  ldapAttrsType =
+    let
+      options = {
+        attrs = mkOption {
+          type = types.attrsOf ldapValueType;
+          default = {};
+          description = "Attributes of the parent entry.";
+        };
+        children = mkOption {
+          # Hide the child attributes, to avoid infinite recursion in e.g. documentation
+          # Actual Nix evaluation is lazy, so this is not an issue there
+          type = let
+            hiddenOptions = lib.mapAttrs (name: attr: attr // { visible = false; }) options;
+          in types.attrsOf (types.submodule { options = hiddenOptions; });
+          default = {};
+          description = "Child entries of the current entry, with recursively the same structure.";
+          example = lib.literalExample ''
+          {
+            "cn=schema" = {
+              # The attribute used in the DN must be defined
+              attrs = { cn = "schema"; };
+              children = {
+                # This entry's DN is expanded to "cn=foo,cn=schema"
+                "cn=foo" = { ... };
+              };
+              # These includes are inserted after "cn=schema", but before "cn=foo,cn=schema"
+              includes = [ ... ];
+            };
+          }
+        '';
+        };
+        includes = mkOption {
+          type = types.listOf types.path;
+          default = [];
+          description = ''
+          LDIF files to include after the parent's attributes but before its children.
+        '';
+        };
+      };
+    in types.submodule { inherit options; };
+
+  valueToLdif = attr: values: let
+    singleValueToLdif = value: if lib.isAttrs value then "${attr}:< file://${value.path}" else "${attr}: ${value}";
```

Comment:
I wonder if that should remove trailing newlines from value?
I had some in my strings and in broke the object. Otherwise there could be a check that `value` does not ends with `\n`.

Reply:
> Continuation Line - A line that begins with a SPACE denotes that the characters following the space are part of the previous line.

Based on this, would replacing all newlines with newline + space be reasonable? This does not preserve the newline in the LDAP value, which may be surprising.

To have a value containing a newline (or other special characters), the user will have to base64-encode the value, I have added another type to allow the user to provide this. Alternatively, we could base64-encode all values - this would be safe, but make the generated `.ldif` much harder to read.

Reply:
Maybe base64 encode if it contains characters that could break encoding otherwise? However there is no builtin for that as far as I can tell.

Reply:
If we're going to dump to a file, we may as well use the `path` version, but it makes the resulting LDIF unreadable. For now, I'm going to leave this up to users to manage I think. I have added support for specifying base64 values.

I have addressed the newline case, by replacing `\n` with `\n<space>`, which evaluates to `<space>` in the attribute. This does stop newlines in values from breaking the LDIF, but might cause surprises? I don't think many values are whitespace-sensitive.